### PR TITLE
HPCC-18658 Refactor ESP sessionID and digital signature data types

### PR DESCRIPTION
**Pull request by RussWhitehead**
Currently ESP stores a user's digital signature and ESP Session ID in
MemoryBuffers. The ESP Session ID is really just an unsigned value,
and should be stored and returned as such. The digital signature will
be a base 64 encoded string, and should be passed as const char *.
This PR makes those changes

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

File: system/security/shared/SecureUser.hpp [hunk 1]
```diff
@@ -214,22 +214,22 @@ class CSecureUser : implements ISecUser, implements ISecCredentials, public CInt
...
-    void setSignature(const MemoryBuffer * const signature)
+    void setSignature(const char * const signature)
     {
         m_signature.clear().append(signature);
     }
 
-    const MemoryBuffer & getSignature()
+    const char * getSignature()
     {
         return m_signature;
```
Comment (wangkx) [inline system/security/shared/SecureUser.hpp]: m_signature.str()?

**Review by wangkx — COMMENTED** [review]
@RussWhitehead 3 minors

File: dali/base/dasess.cpp [hunk 2]
```diff
@@ -2071,13 +2071,13 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
...
             if (len)
-                sessionToken.append(len, mb.readDirect(len));
+                mb.readDirect(sessionToken);
 
             if (mb.remaining() > 0)
             {
                 mb.read(len);
                 if (len)
-                    signature.append(len, mb.readDirect(len));
+                    signature.append(len, (const char *)mb.readDirect(len));
```
[the remark this answers is not in the document]
Comment (RussWhitehead) [reply dali/base/dasess.cpp]: There is no set function that allows you to specify the length

**Comment by RussWhitehead** [comment]
@wangkx Please rereview

File: dali/base/dasess.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (wangkx) [reply dali/base/dasess.cpp]: Do you need to clear before append?

**Comment by RussWhitehead** [comment]
No, because it's a new "udesc" created at the beginning of this case statement

**Comment by wangkx** [comment]
@RussWhitehead looks fine now.

**Comment by RussWhitehead** [comment]
@ghalliday Please merge

**Comment by AttilaVamos** [comment]
It seems something wrong with this PR. Please wait for the Smoketest result.

**Comment by AttilaVamos** [comment]
Excerpt form ESP server log:
```
0000008F 2017-11-03 15:14:59.534 17743 17743 "ESP server started."
00000090 2017-11-03 15:15:58.871 17743 17769 "SYS: LPT=132 APT=79 PU=  3% MU= 11% MAL=132448256 MMP=100024320 SBK=32423936 TOT=521828K RAM=1867160K SWP=67940K"
00000091 2017-11-03 15:16:58.879 17743 17769 "SYS: LPT=132 APT=79 PU=  1% MU= 11% MAL=132448256 MMP=100024320 SBK=32423936 TOT=521828K RAM=1867040K SWP=67940K"
00000092 2017-11-03 15:16:58.879 17743 17769 "DSK: [sda] r/s=0.1 kr/s=3.5 w/s=0.4 kw/s=2.3 bsy=0 NIC: rxp/s=71.4 rxk/s=0.0 txp/s=8.0 txk/s=0.0 CPU: usr=0 sys=0 iow=0 idle=98"
00000093 2017-11-03 15:17:31.639 17743 17911 "ERROR: 1: dasubs.cpp(596) : CDaliPublisherClient::run : Connection to Dali server lost or server closed"
00000094 2017-11-03 15:17:31.708 17743 17743 "Lost Dali Connection"
00000095 2017-11-03 15:17:31.708 17743 17743 "Exiting ESP -- Lost DALI connection!"
00000096 2017-11-03 15:17:31.711 17743 17743 "select handler stopped."
00000097 2017-11-03 15:17:31.713 17743 17743 "ERROR: 1: dasubs.cpp(506) : Dali CDaliSubscriptionManagerStub::remove : Connection to Dali server lost or server closed"
00000098 2017-11-03 15:17:31.713 17743 17743 "ERROR: 1: dasubs.cpp(506) : Dali CDaliSubscriptionManagerStub::remove : Connection to Dali server lost or server closed"
00000099 2017-11-03 15:17:31.715 17743 17743 "ERROR: 1: dasubs.cpp(506) : Dali CDaliSubscriptionManagerStub::remove : Connection to Dali server lost or server closed"
0000009A 2017-11-03 15:17:31.715 17743 17743 "ERROR: 1: dasubs.cpp(506) : Dali CDaliSubscriptionManagerStub::remove : Connection to Dali server lost or server closed"
0000009B 2017-11-03 15:17:31.715 17743 17743 "ERROR: 1: dasubs.cpp(506) : Dali CDaliSubscriptionManagerStub::remove : Connection to Dali server lost or server closed"
0000009C 2017-11-03 15:17:31.715 17743 17743 "QueryFilesInUse.unsubscribe() called"
00000002 2017-11-03 15:17:36.784 23542 23542 "Esp starting community_6.5.0-trunk0[heads/master-0-gd4fc1c]"
00000003 2017-11-03 15:17:36.784 23542 23542 "componentfiles are under /opt/HPCCSystems/componentfiles/"
00000004 2017-11-03 15:17:36.784 23542 23542 "ESP process name [myesp]"
```

**Comment by RussWhitehead** [comment]
@AttilaVamos Can you rerun?  All my PR did was change some variable types in the ISecUser, and is not related to ESP/Dali connections

**Comment by AttilaVamos** [comment]
Rescheduled.

However I executed it on my local Smoketest system and I can see same problem. 
I think that was lethal to change sessionToken from MemoryBuffer to unsigned.

To reproduce:
1. With unit tests (build with -DUSE_CPPUNIT=1 )
  a. Start the system
  b execute:  /opt/HPCCSystems/bin/unittests -e CDaliDFSTests

   the result is 
```
!!!FAILURES!!!
Test Results:
Run:  11   Failures: 0   Errors: 9
```
and core file in mydali, myeclccserver directories

2. With regression tests
  a. Start the system and go to testing/regress/ directory
  b. execute : ./ecl-test setup -t thor

the result is:
```
   -------------------------------------------------
    Result:
    Passing: 3
    Failure: 8
    -------------------------------------------------
    Error:
    setup.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
    
    
    setup_fetch.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
    
    
    setupdict.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
    
    
    setuppersist.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
    
    
    setupsearchindex.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
    
    
    setupsq.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
    
    
    setupwordindex.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
    
    setupxml.ecl
    	Error (1001): Command return code non-zero. 
     retCode: 2
    'Error connecting to 127.0.0.1:8010
    connection failed
    Target: T>127.0.0.1:8010, Raised in: jsocket.cpp, line 1310
    
    SOAP Connection error
```
and core file in  mydali, mythor and myeclccserver directories.

Finally I can see this kind off error messages in dali server log:
```
00000042 2017-11-03 15:55:24.516 28084 28101 "Session starting 800000f0d (192.168.1.58:7215) : role=DfuServer"
00000043 2017-11-03 15:55:32.737 28084 28103 "ERROR: Session Manager - adding unknown session ID 700001792"
00000044 2017-11-03 15:55:32.737 28084 28103 "CServerConnection: connection aborted (80000133a) sessId=700001792"
00000045 2017-11-03 15:55:32.738 28084 28103 "WARNING: 11: dasds.cpp(4595) : EXCEPTION in reply to client 192.168.1.58:7384 : SDS: Connection aborted during connect 
 during connect"
00000046 2017-11-03 15:55:32.739 28084 28085 "Client closed (192.168.1.58:7384)"
00000047 2017-11-03 15:55:35.152 28084 28103 "ERROR: Session Manager - adding unknown session ID 700001368"
00000048 2017-11-03 15:55:35.152 28084 28103 "CServerConnection: connection aborted (800001342) sessId=700001368"
00000049 2017-11-03 15:55:35.153 28084 28103 "WARNING: 11: dasds.cpp(4595) : EXCEPTION in reply to client 192.168.1.58:7338 : SDS: Connection aborted during connect 
 during connect"
0000004A 2017-11-03 15:55:35.155 28084 28103 "WARNING: 18: dasds.cpp(4595) : EXCEPTION in reply to client 192.168.1.58:7338 : SDS: Connection missing (aborted)
 [commit]"
0000004B 2017-11-03 15:55:35.157 28084 28103 "WARNING: 18: dasds.cpp(4595) : EXCEPTION in reply to client 192.168.1.58:7338 : SDS: Connection missing (aborted)
 [commit]"
0000004C 2017-11-03 15:55:35.158 28084 28101 "Session stopping 700001368 ok"
0000004D 2017-11-03 15:55:35.161 28084 28085 "Client closed (192.168.1.58:7338)"
0000004E 2017-11-03 15:55:36.219 28084 28101 "Session starting 800001344 (192.168.1.58:7459) : role=AgentExec"
0000004F 2017-11-03 15:55:38.984 28084 28101 "Session starting 80000176e (192.168.1.58:7170) : role=EclCCServer"
00000050 2017-11-03 15:55:40.649 28084 28089 "SYS: LPT=32 APT=1790 PU= 12% MU= 14% MAL=261410816 MMP=257835008 SBK=3575808 TOT=712956K RAM=2415104K SWP=65940K"
00000051 2017-11-03 15:55:40.650 28084 28089 "DSK: [sda] r/s=3.0 kr/s=26.1 w/s=2.4 kw/s=486.9 bsy=2 NIC: rxp/s=88.3 rxk/s=0.0 txp/s=13.1 txk/s=0.0 CPU: usr=10 sys=1 iow=0 idle=87"
00000052 2017-11-03 15:55:40.650 28084 28089 "KERN_INFO: [17312.695860] eclccserver[27854]: segfault at 7f57ed49adb8 ip 00007f57ed49adb8 sp 00007ffe0f9d2178 error 15 in libc-2.23.so[7f57ed49a000+2000]"
00000053 2017-11-03 15:55:41.065 28084 28101 "Session starting 800001c44 (192.168.1.58:7493) : role=EclAgent"
00000054 2017-11-03 15:55:41.197 28084 28101 "================================================"
00000055 2017-11-03 15:55:41.197 28084 28101 "Program:   192.168.1.58:/opt/HPCCSystems/bin/daserver"
00000056 2017-11-03 15:55:41.197 28084 28101 "Signal:    11 Segmentation fault"
00000057 2017-11-03 15:55:41.197 28084 28101 "Fault IP:  00007F989A7CBA26"
00000058 2017-11-03 15:55:41.197 28084 28101 "Accessing: 00007F98AE303302"
00000059 2017-11-03 15:55:41.197 28084 28101 "Backtrace:"
0000005A 2017-11-03 15:55:41.198 28084 28101 "  /lib/x86_64-linux-gnu/libc.so.6(+0x9fa26) [0x7f989a7cba26]"
0000005B 2017-11-03 15:55:41.198 28084 28101 "  /opt/HPCCSystems/lib/libjlib.so(_ZN12MemoryBuffer10readEndianEjPv+0x65) [0x7f989b321cb5]"
0000005C 2017-11-03 15:55:41.198 28084 28101 "  /opt/HPCCSystems/lib/libdalibase.so(+0xe6532) [0x7f989bc05532]"
0000005D 2017-11-03 15:55:41.198 28084 28101 "  /opt/HPCCSystems/lib/libdalibase.so(+0xdf035) [0x7f989bbfe035]"
0000005E 2017-11-03 15:55:41.198 28084 28101 "  /opt/HPCCSystems/lib/libjlib.so(+0x14c632) [0x7f989b3f7632]"
0000005F 2017-11-03 15:55:41.198 28084 28101 "  /opt/HPCCSystems/lib/libjlib.so(_ZN6Thread5beginEv+0x28) [0x7f989b3ef828]"
00000060 2017-11-03 15:55:41.198 28084 28101 "  /opt/HPCCSystems/lib/libjlib.so(_ZN6Thread11_threadmainEPv+0x1f) [0x7f989b3f144f]"
00000061 2017-11-03 15:55:41.198 28084 28101 "  /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7f989aafd6ba]"
00000062 2017-11-03 15:55:41.198 28084 28101 "  /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f989a8333dd]"
00000063 2017-11-03 15:55:41.198 28084 28101 "Registers:"
00000064 2017-11-03 15:55:41.198 28084 28101 "EAX:000000003E304DF6  EBX:00007F9880000AC8  ECX:0000000000000008  EDX:0000000000000004  ESI:00007F98AE303302  EDI:00007F986FFFE508"
00000065 2017-11-03 15:55:41.198 28084 28101 "R8 :000000000000EFFF  R9 :00000000FFFFFFFF  R10:00000000000002BD  R11:00007F989BBFCFF0"
00000066 2017-11-03 15:55:41.198 28084 28101 "R12:0000000001AAF8D0  R13:00007F9878001120  R14:00007F986FFFE508  R15:00007F9880000B10"
00000067 2017-11-03 15:55:41.198 28084 28101 "CS:EIP:0033:00007F989A7CBA26"
00000068 2017-11-03 15:55:41.198 28084 28101 "   ESP:00007F986FFFE4C8  EBP:0000000000000004"
00000069 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE4C8]: 00007F989B321CB5 01A9DD0000007F98 0000000001A9DD00 80000AC800000000 00007F9880000AC8 01A9DD0000007F98 0000000001A9DD00 9BC0553200000000"
0000006A 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE4E8]: 00007F989BC05532 0000000000007F98 0000000000000000 0000000000000000 0000000700000000 0000000700000007 0000000900000007 0000000400000009"
0000006B 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE508]: 0000000800000004 78000D4000000008 00007F9878000D40 0000000000007F98 0000000000000000 78000B7000000000 00007F9878000B70 0000000000007F98"
0000006C 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE528]: 0000000000000000 0000000000000000 0000000000000000 80000AC800000000 00007F9880000AC8 80000AC800007F98 00007F9880000AC8 0000000100007F98"
0000006D 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE548]: 0000000000000001 0000000000000000 0000000000000000 0000001600000000 0000000000000016 80000B1000000000 00007F9880000B10 9AB0583700007F98"
0000006E 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE568]: 00007F989AB05837 78000A3000007F98 00007F9878000A30 0000000000007F98 0000000000000000 80000BD000000000 00007F9880000BD0 F127340000007F98"
0000006F 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE588]: 9ADC7F10F1273400 6FFFE5B09ADC7F10 00007F986FFFE5B0 8BFFE57000007F98 00007F988BFFE570 80000AC800007F98 00007F9880000AC8 0000000100007F98"
00000070 2017-11-03 15:55:41.198 28084 28101 "Stack[00007F986FFFE5A8]: 0000000000000001 0000000000000000 0000000000000000 0000001700000000 0000000000000017 80000B1000000000 00007F9880000B10 9BBFE03500007F98"
00000071 2017-11-03 15:55:41.198 28084 28101 "ThreadList:
```

**Comment by RussWhitehead** [comment]
@wangkx Please review the minor change to serialization of session Token to Dali

**Comment by wangkx** [comment]
@RussWhitehead looks fine.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Comment by richardkchapman** [comment]
@jakesmith Can you check this too please. I' a bit concerned about "@wangkx Please review the minor change to serialization of session Token to Dali" - does it create any compatibility issues?

**Comment by RussWhitehead** [comment]
@richardkchapman @jakesmith The "minor change to serialization of session Token to Dali" is a change to code I added in this same PR

File: system/security/shared/seclib.hpp [hunk 3]
```diff
@@ -142,10 +142,10 @@ interface ISecCredentials : extends IInterface
 {
     virtual bool setPassword(const char * pw) = 0;
     virtual const char * getPassword() = 0;
-    virtual void setSessionToken(const MemoryBuffer * const token) = 0;
-    virtual const MemoryBuffer & getSessionToken() = 0;
-    virtual void setSignature(const MemoryBuffer * const signature) = 0;
-    virtual const MemoryBuffer & getSignature() = 0;
+    virtual void setSessionToken(unsigned token) = 0;
+    virtual unsigned getSessionToken() = 0;
+    virtual void setSignature(const char * const signature) = 0;
```
Comment (jakesmith) [inline system/security/shared/seclib.hpp]: unusual to make const char * a const param, no real benefit.

File: dali/base/dasess.cpp [hunk 4]
```diff
@@ -2068,16 +2068,13 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
...
-                sessionToken.append(len, mb.readDirect(len));
-
+            mb.read((unsigned &)sessionToken);
             if (mb.remaining() > 0)
             {
+                unsigned len = 0;
                 mb.read(len);
                 if (len)
-                    signature.append(len, mb.readDirect(len));
+                    signature.append(len, (const char *)mb.readDirect(len));
```
Comment (jakesmith) [inline dali/base/dasess.cpp]: > There is no set function that allows you to specify the length

MemoryBuffer::setBuffer

**Comment by jakesmith** [comment]
@RussWhitehead - I think this serialization/deserialization change breaks backward compatibility.

This info is being serialized to Dali - and will not be understood by an old Dali, i.e. it will likely crash.
Or an old client to a new Dali will similarly break.

File: dali/base/dasess.cpp (hunk 4 above)
Comment (RussWhitehead) [reply dali/base/dasess.cpp]: Jake, I am appending this value from the input MemoryBuffer onto the tail end of another MemoryBuffer, so setBuffer will not work

File: system/security/shared/seclib.hpp (hunk 3 above)
Comment (RussWhitehead) [reply system/security/shared/seclib.hpp]: Right you are Jake

**Pull request closed without merging**